### PR TITLE
Remove global access in fv3core and physics tests

### PR DESCRIPTION
**Pull request by elynnwu**
This PR is a prerequisite to moving namelist to driver. The goal for this PR is to 1) remove global config access, 2) share translate class infrastructure, 3) remove dependencies to `fv3core` in physics tests.

Changes:
- Move `grid`, `global_config`, and `corners` from `fv3core.utils` to `pace.util.testing` and `pace.util`: global_config is removed from most places and can be deleted soon-ish
- Move `translate` and `parallel_translate` to `pace.util.testing`: they are shared in dycore and physics
- Move `NamelistDefaults` to `pace.util.namelist`
- Add `PhysicsConfig` class: when namelist is moved to the driver, we should add a method to return physics config
- Add filter in physics savepoint test to only check for physics tests: since the logic checks for every serialized savepoint, pytest also tries to run dycore tests (with no matching test in the physics folder). Adding this avoid checking all serialized savepoints, but future physics test needs to be included here.
- Pass `namelist` and `stencil_factory` down to each savepoint class
- Remove `stencil_facotry` from Grid class

### CODE REVIEW

File: fv3gfs-physics/fv3gfs/physics/testing/translate.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+                self.data[key] = utils.make_storage_data(
+                    value,
+                    shape,
+                    origin=origin,
+                    start=origin,
+                    read_only=True,
+                    backend=self.backend,
+                )
+
+    def python_grid(self):
```
Comment (elynnwu) [inline fv3gfs-physics/fv3gfs/physics/testing/translate.py:350]: All `fv3core` import in physics test have been removed except for `Grid` here. Should we make a new classmethod in `GridData` that makes the class with serialized data? Or should we retain `Grid`? We also need a location for either of them.

File: fv3gfs-physics/fv3gfs/physics/_config.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        if isinstance(value, dict):
+            for subkey, subvalue in value.items():
+                if subkey in flatter_namelist:
+                    raise ValueError(
+                        "Cannot flatten this namelist, duplicate keys: " + subkey
+                    )
+                flatter_namelist[subkey] = subvalue
+        else:
+            flatter_namelist[key] = value
+    return flatter_namelist
```
Comment (mcgibbon) [inline fv3gfs-physics/fv3gfs/physics/_config.py:125]: `namelist_to_flatish_dict` should probably go to `util` since it's used identically in the dycore. This can also live in `pace.util.namelist`.

File: fv3gfs-physics/fv3gfs/physics/testing/translate.py (hunk 1 above)
Comment (mcgibbon) [reply fv3gfs-physics/fv3gfs/physics/testing/translate.py:350]: Grid's main responsibility is to convert serialized data, which may or may not include some number of halos and may or may not be staggered, into standardized storages that all have 3 halos and a compute origin at (3, 3). For this reason, Grid should continue to be used in the test code, and should probably live with the translate infrastructure.

File: fv3gfs-physics/tests/savepoint/translate/translate_fv_update_phys.py [hunk 3]
```diff
@@ -34,9 +33,11 @@ def __getitem__(self, item):
 
 
 class TranslateFVUpdatePhys(ParallelTranslate2Py):
-    def __init__(self, grids):
-        super().__init__(grids)
+    def __init__(self, grids, namelist, stencil_factory):
+        super().__init__(grids, namelist, stencil_factory)
```
Comment (mcgibbon) [inline fv3gfs-physics/tests/savepoint/translate/translate_fv_update_phys.py:37]: Technically it shouldn't be necessary to pass `stencil_factory` separate from `grids`, since each `grid` object has its own stencil factory. It would be different if we are able to remove `grid`, it also would be different if we could remove the `Grid.stencil_factory` attribute and _only_ make stencil factory directly.

Do you think for now we should remove the stencil_factory argument? I can see keeping it if you think this is a step on a road towards retrieving fewer attributes from Grid and eventually being able to use it only for array transformations. I'm kind of undecided on this one.

File: fv3gfs-physics/tests/savepoint/conftest.py [hunk 4]
```diff
@@ -48,8 +57,7 @@ def data_path(pytestconfig):
 def data_path_from_config(config):
     data_path = config.getoption("data_path")
     namelist_filename = os.path.join(data_path, "input.nml")
-    fv3core._config.set_namelist(namelist_filename)
-    return data_path
+    return data_path, namelist_filename
```
Comment (mcgibbon) [inline fv3gfs-physics/tests/savepoint/conftest.py:60]: Please update the function name, and possibly also add a return value type hint.

File: fv3gfs-physics/tests/savepoint/conftest.py [hunk 5]
```diff
@@ -99,12 +111,32 @@ def read_serialized_data(serializer, savepoint, variable):
     return data
 
 
-def process_grid_savepoint(serializer, grid_savepoint, rank, *, backend: str):
-    grid = make_grid(grid_savepoint, serializer, rank, backend=backend)
-    fv3core._config.set_grid(grid)
+def Grid(serializer, grid_savepoint, rank, layout, *, backend: str):
```
Comment (mcgibbon) [inline fv3gfs-physics/tests/savepoint/conftest.py:114]: Please use a lowercase name, since this isn't returning a class.

File: fv3gfs-physics/tests/savepoint/conftest.py [hunk 6]
```diff
@@ -213,16 +264,18 @@ def sequential_savepoint_cases(metafunc, data_path, *, backend: str):
...
 
     # Set the grid to rank 0's data
     serializer = get_serializer(data_path, 0)
     grid_savepoint = serializer.get_savepoint(GRID_SAVEPOINT_NAME)[0]
-    grid_rank0 = process_grid_savepoint(serializer, grid_savepoint, 0, backend=backend)
-    fv3core._config.set_grid(grid_rank0)
-
+    grid_rank0 = Grid(
+        serializer, grid_savepoint, 0, physics_config.layout, backend=backend
+    )
```
Comment (mcgibbon) [inline fv3gfs-physics/tests/savepoint/conftest.py:278]: This variable is unused, you should be able to delete it.

**Comment by mcgibbon** [comment]
I’m a little worried that if we duplicate the base translate classes and make small changes, these classes are very complex and may be hard to merge later. Would it be possible to put these in `pace.util.testing` in a way that both physics and dycore continue to use the same code?

File: fv3gfs-physics/fv3gfs/physics/_config.py (hunk 2 above)
Comment (elynnwu) [reply fv3gfs-physics/fv3gfs/physics/_config.py:125]: Moved to `pace.util.namelist`

File: fv3gfs-physics/tests/savepoint/conftest.py (hunk 5 above)
Comment (elynnwu) [reply fv3gfs-physics/tests/savepoint/conftest.py:114]: Changed to `upper_grid`

File: fv3gfs-physics/fv3gfs/physics/testing/translate.py (hunk 1 above)
Comment (elynnwu) [reply fv3gfs-physics/fv3gfs/physics/testing/translate.py:350]: This makes sense, I had thought that we wanted to get rid of `Grid` all together. If the plan is to keep it alive within the translate infrastructure, I'll go ahead and reshuffle some things.

File: fv3gfs-physics/tests/savepoint/translate/translate_fv_update_phys.py (hunk 3 above)
Comment (elynnwu) [reply fv3gfs-physics/tests/savepoint/translate/translate_fv_update_phys.py:37]: Yes, let's remove the `stencil_factory` argument and access it through grid.

File: fv3core/tests/main/test_config.py [hunk 7]
```diff
@@ -17,12 +17,12 @@
...
-    new_backend = "new_backend"
-    assert new_backend != start_backend
-    fv3core.set_backend(new_backend)
-    assert fv3core.get_backend() == new_backend
+# def test_set_backend():
+#     start_backend = fv3core.get_backend()
+#     new_backend = "new_backend"
+#     assert new_backend != start_backend
+#     fv3core.set_backend(new_backend)
+#     assert fv3core.get_backend() == new_backend
```
Comment (rheacangeo) [inline fv3core/tests/main/test_config.py:25]: this can be deleted now?

File: fv3core/tests/main/test_grid.py [hunk 8]
```diff
@@ -262,7 +262,7 @@ def test_domain_compute(
     add: Index3D,
     domain_compute: Index3D,
 ):
-    grid = fv3core.utils.grid.GridIndexing(
+    grid = pace.stencils.testing.grid.GridIndexing(
```
Comment (rheacangeo) [inline fv3core/tests/main/test_grid.py:265]: is GridIndexing in pace.dsl or pace.stencils.testing.grid?

**Review by rheacangeo — COMMENTED** [review]
Epic. As discussed, there are some known circular dependencies that come about from this. But we agree, this PR is large enough, and if everything works, we should merge this and make a new story to address those smaller cycles.

File: fv3core/fv3core/stencils/delnflux.py [hunk 9]
```diff
@@ -13,9 +13,9 @@
 )
 
 import pace.dsl.gt4py_utils as utils
-from fv3core.utils.grid import DampingCoefficients, axis_offsets
 from pace.dsl.stencil import StencilFactory, get_stencils_with_varied_bounds
 from pace.dsl.typing import FloatField, FloatFieldIJ, FloatFieldK
+from pace.stencils.testing.grid import DampingCoefficients, axis_offsets
```
Comment (mcgibbon) [inline fv3core/fv3core/stencils/delnflux.py:18]: I see most stencil files import from pace.stencils.testing.grid, is there a way to move whatever they’re using out of testing while keeping our testing-only files there? GridData, axis_offsets, and DampingCoefficients don't really belong in `testing`.

File: stencils/pace/stencils/testing/grid.py [hunk 10]
```diff
@@ -1,16 +1,16 @@
...
 
 import numpy as np
 from gt4py import gtscript
 
-import fv3core.utils.global_config as global_config
 import pace.util
-from fv3core.grid import MetricTerms
 from pace.dsl import gt4py_utils as utils
-from pace.dsl.stencil import GridIndexing, StencilConfig, StencilFactory
+from pace.dsl.stencil import GridIndexing
```
Comment (mcgibbon) [inline stencils/pace/stencils/testing/grid.py:11]: Ideally only Grid should remain in this file, the other things should be moved out of `testing`. If we no longer pass `Grid` to `axis_offsets`, you should be able to modify it to only accept GridIndexing, which will allow you to move it out of this file. If we do still pass `Grid` to `axis_offsets`, you can leave the legacy function in this file and call it from whichever testing code is using it.

File: fv3core/fv3core/testing/translate_fvdynamics.py [hunk 11]
```diff
@@ -270,6 +269,8 @@ def __init__(self, grids, *args, **kwargs):
             self.ignore_near_zero_errors[qvar] = True
         self.ignore_near_zero_errors["q_con"] = True
         self.dycore: Optional[fv_dynamics.DynamicalCore] = None
+        self.stencil_factory = stencil_factory
+        self.namelist = namelist
```
Comment (mcgibbon) [inline fv3core/fv3core/testing/translate_fvdynamics.py:273]: Can you type hint `namelist`, since this is a very overloaded term?

File: fv3core/tests/main/test_config.py (hunk 7 above)
Comment (mcgibbon) [inline fv3core/tests/main/test_config.py:25]:
```suggestion
```

File: fv3core/tests/main/test_dycore_call.py [hunk 12]
```diff
@@ -170,9 +170,9 @@ def error_func(*args, **kwargs):
         raise AssertionError("call not allowed")
 
     mock_grid = unittest.mock.MagicMock()
-    with unittest.mock.patch("fv3core.utils.global_config.get_backend", new=error_func):
+    with unittest.mock.patch("pace.util.global_config.get_backend", new=error_func):
         with unittest.mock.patch(
-            "fv3core.utils.global_config.is_gpu_backend", new=error_func
+            "pace.util.global_config.is_gpu_backend", new=error_func
```
Comment (mcgibbon) [inline fv3core/tests/main/test_dycore_call.py:175]: Make sure there's a Jira task to delete pace.util.global_config, if we're no longer using it.

File: fv3core/tests/savepoint/conftest.py [hunk 13]
```diff
@@ -99,12 +100,24 @@ def read_serialized_data(serializer, savepoint, variable):
     return data
 
 
-def process_grid_savepoint(serializer, grid_savepoint, rank, *, backend: str):
-    grid = make_grid(grid_savepoint, serializer, rank, backend=backend)
-    fv3core._config.set_grid(grid)
+def upper_grid(serializer, grid_savepoint, rank, layout, *, backend: str):
```
Comment (mcgibbon) [inline fv3core/tests/savepoint/conftest.py:103]: What is an "upper" grid? Is there a reason for this function when you can directly call `make_grid`?

File: fv3gfs-physics/fv3gfs/physics/testing/__init__.py [hunk 14]
```diff
@@ -1,2 +0,0 @@
-from . import translate_physics
-from .translate_physics import TranslatePhysicsFortranData2Py
```
Comment (mcgibbon) [inline fv3gfs-physics/fv3gfs/physics/testing/__init__.py:2]: Why were these removed?

File: fv3gfs-physics/tests/savepoint/conftest.py [hunk 15]
```diff
@@ -23,9 +24,18 @@
...
+    "GFSPhysicsDriver",
+    "AtmosPhysDriverStatein",
+    "PrsFV3",
+    "PhiFV3",
+    "Microph",
+    "FillGFS",
+    "PhysUpdateTracers",
+    "PhysUpdatePressureSurfaceWinds",
+    "UpdateDWindsPhys",
+    "FVUpdatePhys",
```
Comment (mcgibbon) [inline fv3gfs-physics/tests/savepoint/conftest.py:37]: Cool to have a list of these!

File: fv3gfs-physics/tests/savepoint/conftest.py [hunk 16]
```diff
@@ -99,12 +112,32 @@ def read_serialized_data(serializer, savepoint, variable):
     return data
 
 
-def process_grid_savepoint(serializer, grid_savepoint, rank, *, backend: str):
-    grid = make_grid(grid_savepoint, serializer, rank, backend=backend)
-    fv3core._config.set_grid(grid)
+def upper_grid(serializer, grid_savepoint, rank, layout, *, backend: str):
+    grid = make_grid(grid_savepoint, serializer, rank, layout, backend=backend)
```
Comment (mcgibbon) [inline fv3gfs-physics/tests/savepoint/conftest.py:116]: Can you call `make_grid` directly?

I'm a little confused by this network of functions (`grid_data`, `grid_indexing`, `physics_config`, etc.). Instead of calling `grid_data(grid)`, should you directly use `grid.grid_data` in whatever function is calling these?

**Review by mcgibbon — COMMENTED** [review]
Very nice seeing the test code make direct use of `namelist` and `stencil_factory` instead of relying on `Grid` for everything.

You can leave refactoring `pace.stencils.testing.grid` for a follow-on PR, but please make a Jira task. It's not great having every stencil file import from `testing`, since the idea of putting things there is to indicate the stencil code shouldn't be importing from it.

Very few comments for how much you've updated. Great job!

File: fv3core/examples/standalone/runfile/dynamics.py [hunk 17]
```diff
@@ -227,32 +231,36 @@ def collect_data_and_write_to_file(
             profiler = cProfile.Profile()
             profiler.disable()
 
-        fv3core.set_backend(args.backend)
-        fv3core.set_rebuild(False)
-        fv3core.set_validate_args(False)
```
[the remark this answers is not in the document]
Comment (elynnwu) [reply fv3core/examples/standalone/runfile/dynamics.py:232]: great catch, changed to False

File: fv3core/examples/standalone/runfile/dynamics.py [hunk 18]
```diff
@@ -227,32 +231,36 @@ def collect_data_and_write_to_file(
...
 
         # TODO remove this creation of the legacy grid once everything that
         # references it is updated or removed
-        grid = spec.make_grid_from_namelist(namelist, rank)
-
+        grid = spec.make_grid_from_namelist(dycore_config, rank, args.backend)
+        stencil_config = pace.dsl.stencil.StencilConfig(
+            backend=args.backend,
+            rebuild=False,
+            validate_args=True,
```
Comment (rheacangeo) [inline fv3core/examples/standalone/runfile/dynamics.py:254]: change to False please

File: fv3core/examples/standalone/runfile/dynamics.py [hunk 19]
```diff
@@ -227,32 +231,36 @@ def collect_data_and_write_to_file(
...
+        partitioner = util.CubedSpherePartitioner(
+            util.TilePartitioner(dycore_config.layout)
+        )
         communicator = util.CubedSphereCommunicator(mpi_comm, partitioner)
 
         # TODO remove this creation of the legacy grid once everything that
         # references it is updated or removed
-        grid = spec.make_grid_from_namelist(namelist, rank)
-
+        grid = spec.make_grid_from_namelist(dycore_config, rank, args.backend)
```
[the remark this answers is not in the document]
Comment (elynnwu) [reply fv3core/examples/standalone/runfile/dynamics.py:250]: `stencil_factory` is needed for the baroclinic test case too, so can't be inside `read_serialized_initial_state` only

File: fv3core/tests/main/test_grid.py (hunk 8 above)
Comment (elynnwu) [reply fv3core/tests/main/test_grid.py:265]: It is `pace.dsl`, updated here and elsewhere

File: .pre-commit-config.yaml [hunk 20]
```diff
@@ -33,13 +31,42 @@ repos:
           --follow-imports, skip  # needed so we only test enabled files
         ]
       files: pace-util
+    - id: mypy
```
[the remark this answers is not in the document]
Comment (elynnwu) [reply .pre-commit-config.yaml:34]: Good idea! This fixed one of my problems too, I couldn't figure out it kept running on a file I excluded in stencils, turns out I had 2 mypy stencils and the later one didn't have the exclude

File: examples/dyn_phy.py [hunk 21]
```diff
@@ -2,6 +2,7 @@
 import time
 
 import click
+import f90nml
```
[the remark this answers is not in the document]
Comment (elynnwu) [reply examples/dyn_phy.py:5]: Yes, we should add a plan for this. This script was actually broken for a while and we didn't notice.

File: fv3core/fv3core/_config.py [hunk 22]
```diff
@@ -969,7 +857,7 @@ def set_grid(in_grid):
     grid = in_grid
 
 
-def set_namelist(filename, rank=0):
+def set_namelist(filename, backend, rank=0):
```
[the remark this answers is not in the document]
Comment (elynnwu) [reply fv3core/fv3core/_config.py:860]: Great catch, this one is not called anywhere anymore! Deleting `set_namelist`

File: fv3core/tests/main/test_config.py (hunk 7 above)
Comment (elynnwu) [reply fv3core/tests/main/test_config.py:25]: Rhea noticed the same thing too 👍

File: fv3core/tests/main/test_dycore_call.py (hunk 12 above)
Comment (elynnwu) [reply fv3core/tests/main/test_dycore_call.py:175]: It's being used in two spots in the main tests. Once we get rid of those, we can delete `global_config`

File: fv3core/tests/savepoint/conftest.py (hunk 13 above)
Comment (elynnwu) [reply fv3core/tests/savepoint/conftest.py:103]: It was originally going to be a pytest fixture, but you're right we can just call make_grid directly. Changing this.

File: fv3gfs-physics/tests/savepoint/conftest.py (hunk 16 above)
Comment (elynnwu) [reply fv3gfs-physics/tests/savepoint/conftest.py:116]: They were supposed to be pytest fixtures, but it's easier to make them instead. I forgot to clean up, deleting all these now.

File: fv3gfs-physics/fv3gfs/physics/testing/__init__.py (hunk 14 above)
Comment (elynnwu) [reply fv3gfs-physics/fv3gfs/physics/testing/__init__.py:2]: Since we no longer need to carry separate translate class infrastructure, we don't need this in physics anymore. If we decide to add `TranslatePhysics`, it will go here.

**Review by rheacangeo — APPROVED** [review]
Wahoo! I think the Grid object served us well, but good to finalize the shift away from it by untangling the objects we were grabbing from it as an interim step. Lots going on here and this will make many things possible!

**Review by mcgibbon — APPROVED** [review]
Make sure there are the 2 Jira tasks for removing global_config and for moving utilities out of `pace.stencils.testing.grid`, otherwise looks good to me!
